Alex/implement email notification functionality

alexappleget opened the pull request:
Closes #463 

Create functionality for email notifications into the project for the admin notifications page.

ONCE this ticket gets merged in, I will create more tickets to implement more customization for sending emails. Ex: survivors only email, eliminated. For now, only participants is created to test that the email function does take in id's and sends an email.

1. Used Appwrite's Messaging in order to be able to send email notifications.
2. In order to use Appwrite Messaging we need an email provider that's compatible with Appwrite.
3. I went with Mailgun as the email provider because they have a free version for us for testing.
4. I connected Mailgun to Appwrite Messaging in order to use createEmail() for our code.
5. createEmail() from Appwrite is a backend only function, so I had to create a 'use server' action file to build out the client and create a function to use the createEmail() function inside of it.
6. Since createEmail() is a server function, I used form submission in order to get it to work on the frontend the way we want.
7. Did seperate testing for the client side and then for the server side because the server and client can't crossover.
8. Had to mock the server function inside the client testing because jest isn't configured to parse it.
9. All tests pass and the inputs send an email.


https://github.com/user-attachments/assets/33595540-3ab1-4c3f-899c-a20caf9152a8

ryanfurrer committed:
feat((admin)/layout.tsx): began building out admin layout

ryanfurrer committed:
create admin layout

ryanfurrer committed:
Merge branch 'develop' into ryan/create-admin-layout

ryanfurrer committed:
feat(admin/layout.tsx): added classnames to future components and sorted imports alphabetically

ryanfurrer committed:
Merge branch 'ryan/create-admin-layout' of https://github.com/LetsGetTechnical/gridiron-survivor into ryan/create-admin-layout

ryanfurrer committed:
(admin/layout.tsx): use alias import for errorboundary rather than relative

ryanfurrer committed:
refine admin layout and begin componetizing the layout elements

ryanfurrer committed:
refine components and pages for admin route

ryanfurrer committed:
Merge branch 'develop' into ryan/create-admin-layout

ryanfurrer committed:
attempt to add adminheader props to admin layout so the header can be declared in the layout but the props can be updated per page

ryanfurrer committed:
remove props from adminheader so the layout itself can be reviewed

ryanfurrer committed:
Merge branch 'develop' into ryan/create-admin-layout

alexappleget committed:
Feat: added in email notification functionality

chris-nowicki committed:
fix: removed unused optional props after users in the server action

alexappleget committed:
Auto stash before merge of "alex/implement-email-notification-functionality" and "origin/alex/implement-email-notification-functionality"

Fix: Deleted userIDs from function to avoid getting spammed

alexappleget committed:
Fix: Added in comments to fix eslint error for sendEmailNotification()

alexappleget committed:
Fix: Added participants prop to sendEmailNotification function and created test file

alexappleget committed:
Merge branch 'develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: deleted test file and changed AdminHome to AdminNotifications

alexappleget committed:
Fix: Fixed testing issues for the server function on the client side.

alexappleget committed:
Fix: created test file for sendEmailNotification.ts file to test the function itself

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

ryandotfurrer commented on app/(admin)/admin/notifications/page.test.tsx:
```diff
@@ -1,15 +1,47 @@
     render(<AdminNotifications />);
+
+    subjectInput = screen.getByTestId('subject-text');
+    contentInput = screen.getByTestId('content-text');
```
Update this if you take my previous suggestions into account.

ryandotfurrer commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,46 @@
+// Licensed under the MIT License.
+
+'use server';
+import * as sdk from 'node-appwrite';
```
Do you need to import _everything_?

alexappleget replied:
Unfortunately, since 'client' and 'messaging' is being import 'as sdk' this is the only way to import it correctly.

alexappleget commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,52 @@
  */
 const AdminNotifications = (): JSX.Element => {
+  const [content, setContent] = useState<string>('');
+  const participants = ['66bd072b001f6b1f6ac0'];
```
participants is just what I named it. The createEmail function just takes in userID's and that one is mine for testing. I have another ticket where I am doing the group league of 'Email Testers'. I had to just use mine because I had to manually enter my email in mailgun, verify it, and then use it for testing this

alexappleget commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,52 @@
+        <input
+          type="text"
+          placeholder="Enter title for email.."
+          id="title"
```
i didnt want to use 'test-subject' for the data-testid cuz it sounds bad haha. Thank you for this tip. I'll get the id changed to match the htmlFor

alexappleget commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,52 @@
+          onChange={(e) => setSubject(e.target.value)}
+          data-testid="subject-text"
+        />
+        <label htmlFor="message">Content:</label>
```
I used content because the createEmail function's prop is 'content'. So I will change everything to message (even the prop). If you see the prop changed to 'message', this is why

alexappleget committed:
Fix: updated id's and name's to be consistent

alexappleget committed:
Fix: Added in group emailing functionality.

alexappleget committed:
Merge branch 'alex/add-a-league-into-email-notification' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Alphabetized imports.

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.test.tsx:
```diff
@@ -0,0 +1,35 @@
+        const subject = 'This is a test';
+
+        (sendEmailNotifications as jest.Mock).mockImplementation(async ({content, groupEmailTest, subject}) => {
+            await (messaging.createEmail as jest.Mock)('some-id', subject, content, [], groupEmailTest);
```
Is `some-id` a good way to test this?

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.test.tsx:
```diff
@@ -0,0 +1,35 @@
+describe('SendEmailNotification', () => {
+    it('should send email with provided information', async () => {
+        const content = 'Test';
+        const groupEmailTest = ['123456', '12345', '1234'];
```
I wonder if this mock data could be mock emails instead to be more accurate to the actual data that will be used

alexappleget replied:
I can't mock emails because the data that goes through is userID's. So in the test, I just use random numbers to act like userID's

choir27 replied:
I would change the variable name in that case, because this variable name suggests these should be emails.

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.test.tsx:
```diff
@@ -0,0 +1,35 @@
+
+describe('SendEmailNotification', () => {
+    it('should send email with provided information', async () => {
+        const content = 'Test';
```
I wonder if we can make these test values more specific to our application?

alexappleget replied:
What do you mean by this? The content and subject is something the admin can customize themselves on the frontend. The value can be any thing.

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.test.tsx:
```diff
@@ -0,0 +1,35 @@
+            expect.any(String),
+            subject,
+            content,
+            [],
```
Why are we leaving topics as an empty array here?

alexappleget replied:
The way the createEmail function is set up is it has a strict order of parameters that get passed through. If I rearrange or even delete one (the empty array) it doesn't know what "emailGroup" is. I have to tell the function that the array is empty

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,35 @@
+}): Promise<void> => {
+  try {
+    await messaging.createEmail(
+      ID.unique(), // messageId
+      subject, // subject
+      content, // content
+      [], // topics (optional)
+      groupEmailTest, // users (optional)
```
I wonder if these comments are neccessary

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,35 @@
+      groupEmailTest, // users (optional)
+    );
+  } catch (error) {
+    throw new Error('Error Sending Email');
```
Maybe we should add `error` in this string as well, so the user can see the specifics of the error?

choir27 commented on app/(admin)/admin/notifications/page.test.tsx:
```diff
@@ -1,15 +1,63 @@
 // /Users/ryanfurrer/Developer/GitHub/gridiron-survivor/app/(admin)/admin/notifications/page.test.tsx
 
-import { render, screen } from '@testing-library/react';
 import AdminNotifications from './page';
+import { getCurrentLeague } from '@/api/apiFunctions';
+import { fireEvent, render, screen, waitFor } from '@testing-library/react';
 import React from 'react';
+import { sendEmailNotifications } from './actions/sendEmailNotification';
```
Please alphabetize these

alexappleget replied:
They are already alphabetized

alexappleget replied:
Oh wait fireevent comes before g LOL

choir27 commented on app/(admin)/admin/notifications/page.test.tsx:
```diff
@@ -1,15 +1,63 @@
+    emailTestersButton = screen.getByTestId('email-testers');
+    subjectInput = screen.getByTestId('subject-text');
+  });
+  it(`should render it's content`, () => {
```
`should render it's content` it's here is unclear, is there a way we can reword this to make it clearer?

alexappleget replied:
This wasn't my test. I just copy had to add one thing to it. I think it's all green because I deleted it and repasted it

choir27 commented on app/(admin)/admin/notifications/page.test.tsx:
```diff
@@ -1,15 +1,63 @@
     expect(adminNotificationsContent).toBeInTheDocument();
   });
+  it('should call the sendEmailNotifications function with the provided inputs', async () => {
+    const dummyParticipants = ['12345', '1234', '123'];
```
Is there a way to create this dummyParticpants array so it better reflects the actual data that will be used for this application?

alexappleget replied:
The dummyParticupants are fake userID's to test the function with handling ids

choir27 commented on app/(admin)/admin/notifications/page.test.tsx:
```diff
@@ -1,15 +1,63 @@
+    (getCurrentLeague as jest.Mock).mockResolvedValue({ participants: dummyParticipants });
+
+    fireEvent.click(emailTestersButton);
+    fireEvent.change(subjectInput, { target: { value: 'Test Title' } });
+    fireEvent.change(contentInput, { target: { value: 'Test message section.' } });
```
Is there a way to change these test values so they're more specific to our application?

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,71 @@
+   * To grab all users from the league.
+   */
+  const getLeagueData = async (): Promise<void> => {
+    const leagueId = '66c6618900033d179dda';
```
Is there a reason this leagueId is hardcoded here?

alexappleget replied:
I had to hardcode it because this is the first stage. That leagueID is for a new league I created that only has 3 of us with emails verified on mailgun.

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,71 @@
+    setGroupEmailTest(leagueData.participants);
+    try {
+    } catch (error) {
+      throw new Error('Error Fetching Data');
```
Maybe we should add error in this string as well, so the user can see the specifics of the error?

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,71 @@
   return (
     <section data-testid="admin-notifications-content">
-      <p>{`This is where I'd put my notifation dashboard, IF I HAD ONE!`}</p>
+      <Button
```
Can you please post a screenshot of what this looks like?

choir27 commented on app/(admin)/admin/notifications/page.test.tsx:
```diff
@@ -1,15 +1,63 @@
+
+    await waitFor(() => {
+      expect(sendEmailNotifications as jest.Mock).toHaveBeenCalledWith({
+        content: 'Test message section.',
+        groupEmailTest: dummyParticipants,
+        subject: 'Test Title',
```
Is there a way to change these test values so they're more specific to our application?

alexappleget committed:
Fix: Handled PR comments: deleted unnecessary comments in code, realphabetized imports, changed variable name to groupUsers to avoid confusion, added console.errors.

alexappleget committed:
Fix: Changed the wording of the error.

alexappleget committed:
Fix: Deleted param prop that no longer exists.

alexappleget commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,30 @@
+  subject: string;
+}): Promise<void> => {
+  try {
+    await messaging.createEmail(ID.unique(), subject, content, [], groupUsers);
```
We definitely should but that would have to be another ticket to assign to me

alexappleget committed:
Fix: added the code inside the try block of the function.

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Merge branch 'develop' into alex/implement-email-notification-functionality

alexappleget commented:
@shashilo this is a reminder to review this. It's working and sends emails. I can create another ticket after the ui has been updated to continue with email sending preferences to users.

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Edited league id here to my own email test league with just me.

alexappleget committed:
Fix: Changed variable and function names for more readability.

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.test.tsx:
```diff
@@ -0,0 +1,35 @@
+
+describe('SendEmailNotification', () => {
+    it('should send email with provided information', async () => {
+        const content = 'Test';
+        const sendEmailUsers = ['123456', '12345', '1234'];
+        const subject = 'This is a test';
```
Instead of declaring these as variables, could we hard code some of these values into the tests as they are only used twice?

alexappleget replied:
can you eloborate on this? i don't quite understand

choir27 replied:
Sure, so what I mean is you could do the following:

```
        (sendEmailNotifications as jest.Mock).mockImplementation(async ({content, sendEmailUsers, subject}) => {
            await (messaging.createEmail as jest.Mock)('1234567890', 'This is a test' , 'Test', [], sendEmailUsers);
        })
```

and remove `content` and `subject` as variables since we're only using them in this test twice

alexappleget replied:
i wanted to test that the variables are being passed through correctly. this function is given props from the frontend and it's a backend function. I wanted to make sure that it receives the props correctly and passes it through the function.

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Function to send email.
+ * @param props - subject, content.
+ * @param props.content - The email itself.
```
The description here is a bit vague. Does this include the subject, who is receiving the email, and the email text/attachments?

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,36 @@
+ * Function to send email.
+ * @param props - subject, content.
+ * @param props.content - The email itself.
+ * @param props.subject - The header of the email.
```
I feel like we could be a bit more clear here, maybe add the `header section`.

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,36 @@
+      sendEmailUsers,
+    );
+  } catch (error) {
+    console.error('Error Sending Email:', error);
+    throw new Error('Error Sending Email');
```
Remove console error, as throw new Error can be stacked and takes care of this, as discussed in previous PR's from Shashi

choir27 commented on app/(admin)/admin/notifications/actions/sendEmailNotification.ts:
```diff
@@ -0,0 +1,36 @@
+      content,
+      [],
+      sendEmailUsers,
+    );
```
do we need to add an alert to the user that the email has been successfully sent?

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,72 @@
+  const [sendEmailUsers, setSendEmailUsers] = useState<string[]>([]);
+  const [subject, setSubject] = useState<string>('');
+
+  /**
+   * To grab all users from the league.
+   */
```
Based on the code of `participantsEmail()`, this comment is not accurate.

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,72 @@
+      const leagueData = await getCurrentLeague(leagueId);
+      setSendEmailUsers(leagueData.participants);
+    } catch (error) {
+      console.error('Error Sending Email:', error);
```
Remove console error, as throw new Error can be stacked and takes care of this, as discussed in previous PR's from Shashi

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,72 @@
+   */
+  const participantsEmail = async (): Promise<void> => {
+    try {
+      const leagueId = '66f1a8e300102bff03ff';
+      const leagueData = await getCurrentLeague(leagueId);
```
Is it necessary to separate these into two variables? Could we not just make this into one variable?

alexappleget replied:
it has to be hard coded for now like this

alexappleget committed:
Fix: Cleaned up things per Richard's comments.

alexappleget committed:
Fix: Alphabetized and edited a small change.

choir27 commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,71 @@
+  /**
+   * To grab all participant's userIDs from the league.
+   */
+  const participantsEmail = async (): Promise<void> => {
```
If we're grabbing the participants' userIDs from the league with this function, is calling it `participantsEmail` correct?

alexappleget replied:
yes. the function sends an email to all participants.

alexappleget committed:
Fix: Explained things better in my code.

alexappleget committed:
Merge branch 'develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Added target IDs for bcc

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Fixed testing.

alexappleget committed:
Fix: Fixed tests

alexappleget committed:
Fix: Hard coded bcc target id's

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

shashilo committed:
Merge branch 'develop' into alex/implement-email-notification-functionality

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Merge remote-tracking branch 'origin/alex/implement-email-notification-functionality' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: deleted target ID's

alexappleget committed:
Fix: Fixed tests

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

ryanfurrer committed:
Improved UX for Admin Notification Page (#490)

Closes #474 

Enhance the user experience of the existing admin notification page to
improve its formatting and usability.

**Acceptance Criteria:**
- [X] Review and redesign the layout of the admin notification page for
better organization and clarity.
- [X] Improve the visual hierarchy of notification elements for easier
scanning and management.
- [X] Ensure the design is consistent with the overall admin panel
aesthetics.
---
## Video Demo

https://github.com/user-attachments/assets/cc1e8a0e-45b0-4fbe-9972-2deb085525dc

---------

Co-authored-by: Alex Appleget <alexappleget2014@gmail.com>
Co-authored-by: Shashi Lo <362527+shashilo@users.noreply.github.com>

shashilo committed:
Merge branch 'develop' into alex/implement-email-notification-functionality

ryanfurrer committed:
Merge branch 'develop' into alex/implement-email-notification-functionality

ryanfurrer reviewed (changes requested):
I pulled and ran this locally, and while I'm getting a `POST 200`, I do not see any on-screen confirmation that the email was sent

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Changed leagueId to email testing league.

ryanfurrer reviewed (changes requested):
I am unable to test this; every time I navigate to `/admin` it redirects me to my leagues.

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Created an api function to convert userIDs into their targetIDs for bcc emailing.

alexappleget committed:
Fix: Updated the fetch league data function to now be able to fetch the league info and then grab the targetIDs of the users.

alexappleget committed:
Fix: Updated sendEmailNotification function test.

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Fix: Changed file and folder names.

alexappleget committed:
Fix: fixed testing errors and changed errors to throw error to be more accurate.

alexappleget committed:
Fix: changed mock import to fix test.

choir241 commented on components/Textarea/Textarea.tsx:
```diff
@@ -0,0 +1,121 @@
+/**
+ * useAutosizeTextArea hook
+ * @param {UseAutosizeTextAreaProps} props - The props
+ */
```
Was this Textarea.tsx code automatically generated by shadcn?

choir241 commented on components/Textarea/Textarea.tsx:
```diff
@@ -0,0 +1,121 @@
+        textAreaRef.style.height = `${scrollHeight + offsetBorder}px`;
+      }
+    }
+  }, [textAreaRef, triggerAutoSize]);
```
What happens when we remove textAreaRef and/or triggerAutoSize from useEffect dependencies here?

shashilo replied:
If we don't need to re-render the component, this will cause unnecessary re-renders.

choir241 commented on components/Textarea/Textarea.tsx:
```diff
@@ -0,0 +1,121 @@
+  const [init, setInit] = React.useState(true);
+  React.useEffect(() => {
+    const offsetBorder = 2;
+    if (textAreaRef) {
+      if (init) {
+        textAreaRef.style.minHeight = `${minHeight + offsetBorder}px`;
+        if (maxHeight > minHeight) {
+          textAreaRef.style.maxHeight = `${maxHeight}px`;
+        }
```
Are these if statement nestings necessary, or are there ways we can combine some of them into one if conditional?

alexappleget replied:
This would be a question for Ryan. These were his changes made that got merged into my branch after approval.

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

alexappleget committed:
Merge remote-tracking branch 'origin/develop' into alex/implement-email-notification-functionality

shashilo commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,163 @@
+   */
+  const getLeagueData = async (): Promise<void> => {
+    try {
+      const leagueId = '66f1a8e300102bff03ff'; // EMAIL TEST LEAGUE (DO NOT CHANGE)
```
Do we not have this dynamically working?

shashilo commented on app/(admin)/admin/notifications/page.tsx:
```diff
@@ -1,16 +1,163 @@
+     * Fetches the league data.
+     * @returns The league data.
+     */
+    const fetchData = async (): Promise<void> => {
```
Does this need to be in a useEfect or can it just be called when the component mounts?
